---
title: "Splising Sites Selection"
author: "Nadzeya Boyeva"
date: "2024-11-05"
output: pdf_document
---

```{r include=FALSE}
library(dplyr)
library(edgeR)
library(Rsamtools)
library(ggplot2)
library(ggseqlogo)
source("3_SNVs_in_splicing_sites_functions.R")
#source("/media/user5/new/boyeva/r/2_splicing_sites_selection_functions.R")
```

# Sites from EEJ gene related matrices

```{r}
fd <- '/home/nadzeya/praktika'
fd116 <- '/media/user5/new/boyeva'

readTxt <- function(path) {
  read.table(file=path, 
             sep="\t",
             header=TRUE,
             quote="\"",
             as.is=TRUE)
}
```

EEJ gene related matrices contain information about exon-exon junctions which is inferred from RNAseq reads mapping gaps. Rows of every matrix of this kind are individual EEJs, and columns are samples. To create a set of EEJs, where splicing actually occurs in Kasumi-1 cells, we need to combine all the EEJ counts from all the samples:

```{r}
eej_dataset_dir <- file.path(fd116, "Dataset_of_EEJs")
all_files <- list.files(eej_dataset_dir, full.names = TRUE)
combined_df <- readTxt(all_files[[1]])

for (i in 2:length(all_files)) {
  combined_df <- full_join(combined_df, 
                           readTxt(all_files[[i]]), by = c("eej_id", 
                                                           "gene_id", 
                                                           "seqnames", 
                                                           "start", 
                                                           "end", 
                                                           "width", 
                                                           "strand"), copy=TRUE)
}

combined_df[is.na(combined_df)] <- 0

nrow(combined_df)
ncol(combined_df)
colnames(combined_df)[1:10]
```

To keep all potentially active EEJs , firstly, we selected low filtering threshold: an EEJ should have at leat 10 reads supporting it or CPM \>= -1 (0.5 per million reads) at least in one sample.

```{r}
filtered_df <- filter_eej_df(combined_df,
                              use_raw=TRUE,
                              use_cpm=TRUE,
                              raw_threshold=10, 
                              cpm_threshold=-1,
                              n_samples_threshold=1)

filtered_df$mean_raw <- filtered_df$sum_raw / filtered_df$sum_samples_raw
filtered_df$mean_cpm <- filtered_df$sum_cpm / filtered_df$sum_samples_cpm

write.table(filtered_df, file = file.path(fd116, "EEJ_raw10_cpm-1_s1.txt"), sep = '\t')
str(filtered_df)
```

We repeated filtration with higher CPN threshold:

```{r}
filtered_df <- filter_eej_df(combined_df,
                              use_raw=TRUE,
                              use_cpm=TRUE,
                              raw_threshold=10, 
                              cpm_threshold=0,
                              n_samples_threshold=1)

filtered_df$mean_raw <- filtered_df$sum_raw / filtered_df$sum_samples_raw
filtered_df$mean_cpm <- filtered_df$sum_cpm / filtered_df$sum_samples_cpm

write.table(filtered_df, file = file.path(fd116, "EEJ_raw10_cpm0_s1.txt"), sep = '\t')
nrow(filtered_df)
```

... and with higher supporting samples number threshold:

```{r}
filtered_df <- filter_eej_df(combined_df,
                              use_raw=TRUE,
                              use_cpm=TRUE,
                              raw_threshold=10, 
                              cpm_threshold=0,
                              n_samples_threshold=3)

filtered_df$mean_raw <- filtered_df$sum_raw / filtered_df$sum_samples_raw
filtered_df$mean_cpm <- filtered_df$sum_cpm / filtered_df$sum_samples_cpm

write.table(filtered_df, file = file.path(fd116, "EEJ_raw10_cpm0_s3.txt"), sep = '\t')
nrow(filtered_df)
```

# Logos

To explore logos of splicing sites in selected EEJ regions, we import reference genome and extract sequences of splicing sites from it.

```{r}
ref_path = file.path(fd, "Homo_sapiens.GRCh38.dna_sm.toplevel.fa")
ref_idx_path = file.path(fd, "Homo_sapiens.GRCh38.dna_sm.toplevel.fa.fai")
file <- FaFile(ref_path, index=ref_idx_path)
fasta <- open(file)
```

We can see that GT dinucleotide in 5' SS is not that conservative with lower filtering thresholds.

```{r}
input <- "EEJ_raw10_cpm-1_s1.txt"

res <- get_SS_from_EEJ(read_from_file=TRUE, file.path(fd, input))
res$fiveSSs@seqnames <- gsub("chr", "", res$fiveSSs@seqnames)
x1 <- getSeq(x=fasta, res[[1]])
ggseqlogo(data.frame(x1), seq_type='dna', method='prob')
```

We see the same situation for 3' SS dinucleotide â€“ normally it should be more conservative.

```{r}
res$threeSSs@seqnames <- gsub("chr", "", res$threeSSs@seqnames)
x2 <- getSeq(x=fasta, res[[2]])
ggseqlogo(data.frame(x2), seq_type = 'dna', method='prob')
```

But if we take higher thresholds, dinucleotides are more conservative here.

```{r}
input <- "EEJ_raw10_cpm0_s3.txt"

res <- get_SS_from_EEJ(read_from_file=TRUE, file.path(fd, input))
res$fiveSSs@seqnames <- gsub("chr", "", res$fiveSSs@seqnames)
x1 <- getSeq(x=fasta, res[[1]])
ggseqlogo(data.frame(x1), method='prob')
```

```{r}
res$threeSSs@seqnames <- gsub("chr", "", res$threeSSs@seqnames)
x2 <- getSeq(x=fasta, res[[2]])
ggseqlogo(data.frame(x2), seq_type = 'dna', method='prob')
```
