---
title: "Splising Sites Selection"
author: "Nadzeya Boyeva"
date: "2024-11-05"
output: pdf_document
---

```{r}
library(dplyr)
library(edgeR)
library(Rsamtools)
library(ggplot2)
library(ggseqlogo)
source("3_SNVs_in_splicing_sites_functions.R")
#source("/media/user5/new/boyeva/r/2_splicing_sites_selection_functions.R")
```

# Sites from EEJ gene related matrices

```{r}
fd116 <- '/media/user5/new/boyeva'

readTxt <- function(path) {
  read.table(file=path, 
             sep="\t",
             header=TRUE,
             quote="\"",
             as.is=TRUE)
}
```

```{r}
eej_dataset_dir <- file.path(fd116, "Dataset_of_EEJs")
all_files <- list.files(eej_dataset_dir, full.names = TRUE)
combined_df <- readTxt(all_files[[1]])

for (i in 2:length(all_files)) {
  combined_df <- full_join(combined_df, 
                           readTxt(all_files[[i]]), by = c("eej_id", "gene_id", "seqnames", "start", "end", "width", "strand"), copy=TRUE)
}

combined_df[is.na(combined_df)] <- 0

nrow(combined_df)
ncol(combined_df)
colnames(combined_df)[1:10]
```

```{r}
filtered_df <- filter_eej_df(combined_df,
                              use_raw=TRUE,
                              use_cpm=TRUE,
                              raw_threshold=10, 
                              cpm_threshold=-1,
                              n_samples_threshold=1)

filtered_df$mean_raw <- filtered_df$sum_raw / filtered_df$sum_samples_raw
filtered_df$mean_cpm <- filtered_df$sum_cpm / filtered_df$sum_samples_cpm

write.table(filtered_df, file = file.path(fd116, "EEJ_raw10_cpm-1_s1.txt"), sep = '\t')
str(filtered_df)
```

```{r}
filtered_df <- filter_eej_df(combined_df,
                              use_raw=TRUE,
                              use_cpm=TRUE,
                              raw_threshold=10, 
                              cpm_threshold=0,
                              n_samples_threshold=1)

filtered_df$mean_raw <- filtered_df$sum_raw / filtered_df$sum_samples_raw
filtered_df$mean_cpm <- filtered_df$sum_cpm / filtered_df$sum_samples_cpm

write.table(filtered_df, file = file.path(fd116, "EEJ_raw10_cpm0_s1.txt"), sep = '\t')
nrow(filtered_df)
```

```{r}
filtered_df <- filter_eej_df(combined_df,
                              use_raw=TRUE,
                              use_cpm=TRUE,
                              raw_threshold=10, 
                              cpm_threshold=0,
                              n_samples_threshold=3)

filtered_df$mean_raw <- filtered_df$sum_raw / filtered_df$sum_samples_raw
filtered_df$mean_cpm <- filtered_df$sum_cpm / filtered_df$sum_samples_cpm

write.table(filtered_df, file = file.path(fd116, "EEJ_raw10_cpm0_s3.txt"), sep = '\t')
nrow(filtered_df)
```

# Setecting sites within categories

```{r}
kasumi_dir <- file.path(fd, "Kasumi-1_cells")
kasumi_files <- c(list.files(kasumi_dir, full.names = TRUE),
               file.path(fd, "Dataset_of_EEJs", "Kasumi-1, control cells, EEJs from original RNA-Seq.txt"))
```

```{r}
kasumi_df <- readTxt(kasumi_files[[1]])

for (i in 2:length(kasumi_files)) {
  kasumi_df <- full_join(kasumi_df, 
                           readTxt(kasumi_files[[i]]), by = c("eej_id", "gene_id", "seqnames", "start", "end", "width", "strand"))
}

kasumi_df[is.na(kasumi_df)] <- 0
kasumi_df
```

```{r}
kasumi_cpm <- get_eej(kasumi_df)
write.table(kasumi_cpm, file = file.path(fd, "EEJ_Kasumi_filtered_CPM-1.txt"), sep = '\t')

kasumi_raw <- get_eej(kasumi_df, use_raw = TRUE)
write.table(kasumi_cpm, file = file.path(fd, "EEJ_Kasumi_filtered_raw10.txt"), sep = '\t')

common <- (kasumi_cpm$eej_id %in% kasumi_raw$eej_id)  
kasumi_dc <- kasumi_cpm[common,]
write.table(kasumi_dc, file = file.path(fd, "EEJ_Kasumi_filtered_DC.txt"), sep = '\t')
```

```{r}
ref_path = file.path(fd, "Homo_sapiens.GRCh38.dna_sm.toplevel.fa")
ref_idx_path = file.path(fd, "Homo_sapiens.GRCh38.dna_sm.toplevel.fa.fai")
file <- FaFile(ref_path, index=ref_idx_path)
fasta <- open(file)


kasumi_dc$seqnames <- gsub("chr", "", kasumi_dc$seqnames)
kasumi_dc_short <- cbind(kasumi_dc[,1:8], sum_samples = kasumi_dc$sum_samples)
kasumi_dc_ss_coords <- get_SS_from_EEJ(read_from_file=FALSE, df=kasumi_dc_short)

ss <- c(rep(5, length(kasumi_dc_ss_coords$fiveSSs)),
        rep(3, length(kasumi_dc_ss_coords$threeSSs)))

kasumi_dc_ss_coords_df <- rbind(as.data.frame(kasumi_dc_ss_coords$fiveSSs), 
                                as.data.frame(kasumi_dc_ss_coords$threeSSs))

kasumi_dc_ss_coords_df$ss <- ss
kasumi_dc_ss_coords_df$ss_start <- kasumi_dc_ss_coords_df$start
kasumi_dc_ss_coords_df$ss_end <- kasumi_dc_ss_coords_df$end

kasumi_dc_ss_coords_df <- add_refseqs(fasta, kasumi_dc_ss_coords_df, source="EEJ")
write.table(kasumi_dc_ss_coords_df, file = file.path(fd, "EEJ_Kasumi_filtered_DC_with_refseqs.txt"), sep = '\t')
```

# Logos

```{r}
input <- "RUNX1-RUNX1T1 project, EEJs from Kasumi-1 cells, genes-related matrix.txt"

ref_path = file.path(fd, "Homo_sapiens.GRCh38.dna_sm.toplevel.fa")
ref_idx_path = file.path(fd, "Homo_sapiens.GRCh38.dna_sm.toplevel.fa.fai")
file <- FaFile(ref_path, index=ref_idx_path)
fasta <- open(file)

add_refseqs_plot_logo <- function(filepath, ref_genome) {
  res <- get_SS_from_EEJ(read_from_file=TRUE, filepath)
  res$fiveSSs@seqnames <- gsub("chr", "", res$fiveSSs@seqnames)
  res$fiveSSs$refseqs <- getSeq(x=ref_genome, res[[1]])
  ggplot() + geom_logo(res$fiveSSs$refseqs) + theme_logo()
  
  res$threeSSs@seqnames <- gsub("chr", "", res$threeSSs@seqnames)
  res$threeSSs$refseqs <- getSeq(x=ref_genome, res[[1]])
  ggplot() + geom_logo(res$threeSSs$refseqs) + theme_logo()
}
add_refseqs_plot_logo(file.path(fd, "Kasumi-1_cells", input), fasta)
```

```{r}
input <- "EEJ_raw10_cpm0_s3.txt"

ref_path = file.path(fd, "Homo_sapiens.GRCh38.dna_sm.toplevel.fa")
ref_idx_path = file.path(fd, "Homo_sapiens.GRCh38.dna_sm.toplevel.fa.fai")
file <- FaFile(ref_path, index=ref_idx_path)
fasta <- open(file)


res <- get_SS_from_EEJ(read_from_file=TRUE, file.path(fd, input))
res$fiveSSs@seqnames <- gsub("chr", "", res$fiveSSs@seqnames)
x1 <- getSeq(x=fasta, res[[1]])
ggseqlogo(data.frame(x1))
```

```{r}
res$threeSSs@seqnames <- gsub("chr", "", res$threeSSs@seqnames)
x2 <- getSeq(x=fasta, res[[2]])
ggplot() + geom_logo(data.frame(x2)) + theme_logo()
```
