---
title: "SNVs in Splicing SItes"
author: "Nadzeya Boyeva"
date: "2024-11-09"
output: pdf_document
---

# Search of SNVs in Splicing Sites

```{r include=FALSE}
fd <- '/home/nadzeya/praktika'
library(VariantAnnotation)
library(ggplot2)
library(dplyr)
library(ggseqlogo)
source("/home/nadzeya/Documents/GitHub/Splicing_SNV/1_SNV_examination_functions.R")
source("/home/nadzeya/Documents/GitHub/Splicing_SNV/3_SNVs_in_splicing_sites_functions.R")
```

```{r}
readTxt <- function(path) {
  read.table(file=path, 
             sep="\t",
             header=TRUE,
             quote="\"",
             as.is=TRUE)
}
```

Read VCF files from first, second and third nascent RNA samples, and all samples merged into one file on alignment step. Retrieve common SNVs for all four files (nas_common).

```{r}
nas_vcf1_path <- file.path(fd, "nas1_snps_f.vcf.gz")
nas_vcf2_path <- file.path(fd, "nas2_snps_f.vcf.gz")
nas_vcf3_path <- file.path(fd, "nas3_2_snps_f.vcf.gz")
nas_merged_vcf_path <- file.path(fd, "nas_merged_snps_f.vcf.gz")

nas1 <- load_vcf(nas_vcf1_path)
nas2 <- load_vcf(nas_vcf2_path)
nas3 <- load_vcf(nas_vcf3_path)
nas_merged <- load_vcf(nas_merged_vcf_path)

nas_common <- nas_merged[nas_merged@ranges@NAMES %in% 
                           Reduce(intersect, list(nas1@ranges@NAMES, 
                                                  nas2@ranges@NAMES, 
                                                  nas3@ranges@NAMES, 
                                                  nas_merged@ranges@NAMES))]
```

## SSs from RNA-Seq derived EEJ matrices

Load filtered EEJ dataset. We will use dataset where each EEJ has read count \>=10 and log10CPM \>=0 at least in three samples. Then unify chromosome notation, select metadata EEJ columns, and extract SSs coordinates from EEJ coordinates.

```{r}
ss_df <- readTxt(file.path(fd, "EEJ_raw10_cpm0_s3.txt"))
ss_df$seqnames <- gsub("chr", "", ss_df$seqnames)
ss_coords <- get_SS_from_EEJ(read_from_file=FALSE, df=ss_df)
```

Find SNVs overlapping with SSs.

```{r}
eej_nas_common <- find_overlaps_jointSS(ss_coords, nas_common, ss_df)
nrow(eej_nas_common)
```

Add reference and alternative sequences of SSs.

```{r}
ref_path = file.path(fd, "Homo_sapiens.GRCh38.dna_sm.toplevel.fa")
ref_idx_path = file.path(fd, "Homo_sapiens.GRCh38.dna_sm.toplevel.fa.fai")
file <- FaFile(ref_path, index=ref_idx_path)
fasta <- open(file)
```

```{r}
eej_nas_common <- add_refseqs(fasta, eej_nas_common)
eej_nas_common <- add_altseqs(eej_nas_common)
write.table(eej_nas_common, file=file.path(fd, "eej_nas_common.txt"), sep='\t')
head(eej_nas_common)
```

```{r}
ggseqlogo(eej_nas_common$refseq[eej_nas_common$ss == "5"])
```

```{r}
ggseqlogo(eej_nas_common$altseq[eej_nas_common$ss == "5"])
```

```{r}
ggseqlogo(eej_nas_common$refseq[eej_nas_common$ss == "3"])
```

```{r}
ggseqlogo(eej_nas_common$altseq[eej_nas_common$ss == "3"])
```

```{r}
eej_nas_common_5ss <- eej_nas_common[eej_nas_common$ss == 5,]
eej_nas_common_3ss <- eej_nas_common[eej_nas_common$ss == 3,]

# Combine the data into one dataframe with group labels
combined <- rbind(
  data.frame(snp_pos_in_ss = eej_nas_common_5ss$snp_pos_in_ss, group = "5'SS"),
  data.frame(snp_pos_in_ss = eej_nas_common_3ss$snp_pos_in_ss, group = "3'SS")
  )

# Create the ggplot density plot
ggplot(combined, aes(x = snp_pos_in_ss, color = group)) +
  geom_density(linewidth = 1) +  # Add density lines
  labs(
    title = "Density Plot of SNP Positions by Splice Site and Strand",
    x = "SNP Position in Splice Site",
    y = "Density",
    color = "Group"
  ) +
  theme_minimal()
```

```{r}
counts <- combined %>%
  group_by(group, snp_pos_in_ss) %>%
  summarise(count = n(), .groups = "drop")

dinucleotide_counts <- counts %>% # dots on the plot are dinucleotides
  filter(
    (group == "5'SS" & (snp_pos_in_ss == 4 | snp_pos_in_ss == 5)) |
    (group == "3'SS" & (snp_pos_in_ss == 19 | snp_pos_in_ss == 20))
    )

ggplot(combined, aes(x = snp_pos_in_ss, color = group)) +
  geom_freqpoly(binwidth = 1, linewidth = 1) +  # Main frequency lines
  geom_point(
    data = dinucleotide_counts,
    aes(x = snp_pos_in_ss, y = count, color = group),
    size = 4, shape = 19
  ) +
  labs(
    title = "Frequency Plot of SNP Positions with Splice Site Dinucleotides",
    x = "SNP Position in Splice Site",
    y = "Count",
    color = "Group"
  ) +
  theme_minimal()
```

## Unique SSs from UCSC Annotation

```{r}
bed_path = file.path(fd, "ucsc.bed")
ucsc <- import(con = bed_path, format = "BED")
ucsc@seqnames <- gsub("chr", "", ucsc@seqnames)
ucsc <- ucsc[nchar(as.character(ucsc@seqnames)) < 3, ]
ucsc@seqnames <- droplevels(ucsc@seqnames)
ucsc_df <- as.data.frame(ucsc)
ucsc_df <- ucsc_df[!duplicated(ucsc_df[c("seqnames", "start", "end", "strand")]),]

ucsc_df$refseq_id <- sapply(strsplit(ucsc_df$name, "_"), function(x) paste(x[1], x[2], sep = "_"))
ucsc_df$eej_id <- paste0("chr", as.character(ucsc_df$seqnames), ":", 
                            as.character(ucsc_df$start - 1), "-", 
                            as.character(ucsc_df$end + 1), "_str", 
                            ucsc_df$strand)
ucsc_df <- ucsc_df[ucsc_df$eej_id %in%setdiff(ucsc_df$eej_id, ss_df$eej_id),]
rownames(ucsc_df) <- 1:nrow(ucsc_df)
ucsc_df$start <- ucsc_df$start - 1
ucsc_df$end <- ucsc_df$end + 1
ucsc_df
```

```{r}
ss_coords_ucsc <- get_SS_from_EEJ(read_from_file = FALSE, df=ucsc_df)
ucsc_nas_common <- find_overlaps_jointSS(ss_coords_ucsc, nas_common, ucsc_df)
ucsc_nas_common <- add_refseqs(fasta, ucsc_nas_common)
ucsc_nas_common <- add_altseqs(ucsc_nas_common)

ucsc_nas_common$refseq_id <- sapply(strsplit(ucsc_nas_common$name, "_"), function(parts) {
  paste(parts[1], parts[2], sep = "_")
})

write.table(ucsc_nas_common, file=file.path(fd, "ucsc_nas_common.txt"), sep='\t')
ucsc_nas_common
```

```{r}
ucsc_nas_common_5ss_plus <- ucsc_nas_common[(ucsc_nas_common$ss == 5 & ucsc_nas_common$strand == '+'),]
ucsc_nas_common_5ss_minus <- ucsc_nas_common[(ucsc_nas_common$ss == 5 & ucsc_nas_common$strand == '-'),]
ucsc_nas_common_3ss_plus <- ucsc_nas_common[(ucsc_nas_common$ss == 3 & ucsc_nas_common$strand == '+'),]
ucsc_nas_common_3ss_minus <- ucsc_nas_common[(ucsc_nas_common$ss == 3 & ucsc_nas_common$strand == '-'),]

ucsc_combined <- rbind(
  data.frame(snp_pos_in_ss = ucsc_nas_common_5ss_plus$snp_pos_in_ss, group = "5SS +"),
  data.frame(snp_pos_in_ss = ucsc_nas_common_5ss_minus$snp_pos_in_ss, group = "5SS -"),
  data.frame(snp_pos_in_ss = ucsc_nas_common_3ss_plus$snp_pos_in_ss, group = "3SS +"),
  data.frame(snp_pos_in_ss = ucsc_nas_common_3ss_minus$snp_pos_in_ss, group = "3SS -")
)
```

```{r}
ucsc_counts <- ucsc_combined %>%
  group_by(group, snp_pos_in_ss) %>%
  summarise(count = n(), .groups = "drop")

ucsc_dinucleotide_counts <- ucsc_counts %>%
  filter(
    (group == "5SS +" & (snp_pos_in_ss == 3 | snp_pos_in_ss == 4)) |  # GU for 5SS +
    (group == "5SS -" & (snp_pos_in_ss == 4 | snp_pos_in_ss == 5)) |  # GU for 5SS -
    (group == "3SS +" & (snp_pos_in_ss == 18 | snp_pos_in_ss == 19)) | # AG for 3SS +
    (group == "3SS -" & (snp_pos_in_ss == 3 | snp_pos_in_ss == 4))    # AG for 3SS -
  )

ggplot(ucsc_combined, aes(x = snp_pos_in_ss, color = group)) +
  geom_freqpoly(binwidth = 1, linewidth = 1) +  # Main frequency lines
  geom_point(
    data = ucsc_dinucleotide_counts,
    aes(x = snp_pos_in_ss, y = count, color = group),
    size = 4, shape = 19
  ) +
  labs(
    title = "Frequency Plot of SNP Positions with Splice Site Dinucleotides",
    x = "SNP Position in Splice Site",
    y = "Count",
    color = "Group"
  ) +
  theme_minimal()
```

Save SNVs which are in dinucleotides in SSs unique to UCSC source.

Load list of expressed genes (leg):

```{r}
leg <- readTxt(file.path(fd, "RUNX1-RUNX1T1 project, list of expressed genes"))
head(leg)
```

Annotate and save the dataframe:

```{r}
ucsc_nas_common_dinucl <- rbind(
  ucsc_nas_common[ucsc_nas_common$ss == "5" & 
                       ucsc_nas_common$strand == "+" &
                       (ucsc_nas_common$snp_pos_in_ss %in% c("3", "4")),],
  ucsc_nas_common[ucsc_nas_common$ss == "5" & 
                       ucsc_nas_common$strand == "-" &
                       (ucsc_nas_common$snp_pos_in_ss %in% c("4", "5")),],
  ucsc_nas_common[ucsc_nas_common$ss == "3" & 
                       ucsc_nas_common$strand == "+" &
                       (ucsc_nas_common$snp_pos_in_ss %in% c("18", "19")),],
  ucsc_nas_common[ucsc_nas_common$ss == "3" & 
                       ucsc_nas_common$strand == "-" &
                       (ucsc_nas_common$snp_pos_in_ss %in% c("3", "4")),])

leg_dinucl <- leg[match(ucsc_nas_common_dinucl$refseq_id, leg$refseq_id),]
ucsc_nas_common_dinucl$gene_symbol <- leg_dinucl$gene_symbol
ucsc_nas_common_dinucl$gene_name <- leg_dinucl$gene_name

write.table(ucsc_nas_common_dinucl, file=file.path(fd, "ucsc_nas_common_dinucleotides.txt"), sep='\t')
ucsc_nas_common_dinucl
```
