---
title: "SNVs in Splicing SItes"
author: "Nadzeya Boyeva"
date: "2024-11-09"
output: pdf_document
---

```{r include=FALSE}
fd <- '/home/nadzeya/praktika'
library(VariantAnnotation)
library(ggplot2)
source("/home/nadzeya/Documents/GitHub/Splicing_SNV/SNV_examination_functions.R")
source("/home/nadzeya/Documents/GitHub/Splicing_SNV/SNVs_in_splicing_sites_functions.R")
```

```{r}
readTxt <- function(path) {
  read.table(file=path, 
             sep="\t",
             header=TRUE,
             quote="\"",
             as.is=TRUE)
}
```

```{r}
nas_vcf1_path <- file.path(fd, "nas1_snps_f.vcf.gz")
nas_vcf2_path <- file.path(fd, "nas2_snps_f.vcf.gz")
nas_vcf3_path <- file.path(fd, "nas3_2_snps_f.vcf.gz")
nas_merged_vcf_path <- file.path(fd, "nas_merged_snps_f.vcf.gz")

nas1 <- load_vcf(nas_vcf1_path)
nas2 <- load_vcf(nas_vcf2_path)
nas3 <- load_vcf(nas_vcf3_path)
nas_merged <- load_vcf(nas_merged_vcf_path)

nas_common <- nas_merged[nas_merged@ranges@NAMES %in% 
                           Reduce(intersect, list(nas1@ranges@NAMES, 
                                                  nas2@ranges@NAMES, 
                                                  nas3@ranges@NAMES, 
                                                  nas_merged@ranges@NAMES))]
```

```{r}
ss_df <- readTxt(file.path(fd, "EEJ_filtered_DC.txt"))
ss_df$seqnames <- gsub("chr", "", ss_df$seqnames)
ss_df_short <- ss_df[,1:8]
ss_coords <- get_SS_from_EEJ(read_from_file=FALSE, df=ss_df_short)
```

```{r}
dc_nas_common <- find_overlaps_jointSS(ss_coords, nas_common, ss_df_short)
write.table(dc_nas_common, file=file.path(fd, "dc_eej_nas_common.txt"), sep='\t')
dc_nas_common
```

```{r}
#dc_nas_common <- readTxt(file.path(fd, "dc_eej_nas_common.txt"))
dc_nas_common_5ss_plus <- dc_nas_common[(dc_nas_common$ss == 5 & dc_nas_common$strand == '+'),]
dc_nas_common_5ss_minus <- dc_nas_common[(dc_nas_common$ss == 5 & dc_nas_common$strand == '-'),]
dc_nas_common_3ss_plus <- dc_nas_common[(dc_nas_common$ss == 3 & dc_nas_common$strand == '+'),]
dc_nas_common_3ss_minus <- dc_nas_common[(dc_nas_common$ss == 3 & dc_nas_common$strand == '-'),]
```

```{r}
# Combine the data into one dataframe with group labels
dc_combined <- rbind(
  data.frame(snp_pos_in_ss = dc_nas_common_5ss_plus$snp_pos_in_ss, group = "5SS +"),
  data.frame(snp_pos_in_ss = dc_nas_common_5ss_minus$snp_pos_in_ss, group = "5SS -"),
  data.frame(snp_pos_in_ss = dc_nas_common_3ss_plus$snp_pos_in_ss, group = "3SS +"),
  data.frame(snp_pos_in_ss = dc_nas_common_3ss_minus$snp_pos_in_ss, group = "3SS -")
)

# Create the ggplot density plot
ggplot(dc_combined, aes(x = snp_pos_in_ss, color = group)) +
  geom_density(linewidth = 1) +  # Add density lines
  labs(
    title = "Density Plot of SNP Positions by Splice Site and Strand",
    x = "SNP Position in Splice Site",
    y = "Density",
    color = "Group"
  ) +
  theme_minimal()  # Use a clean theme
```

```{r}
dc_nas_merged <- find_overlaps_jointSS(ss_coords, nas_merged, ss_df_short)
write.table(dc_nas_merged, file=file.path(fd, "dc_eej_nas_merged.txt"), sep='\t')
dc_nas_merged
```

```{r}
leg <- readTxt(file.path(fd, "RUNX1-RUNX1T1 project, list of expressed genes"))
leg
```

# Actively expressing genes selection

```{r}
dc_nas_common[dc_nas_common$gene_id %in% leg$gene_id,]
```

```{r}
dc_nas_merged[dc_nas_merged$gene_id %in% leg$gene_id,]
```

# Check normally not-expressing genes with SNVs

```{r}
genes_annotation <- readTxt(file.path(fd, 
                                      "Ensembl release 85, GRCh38.p7, annotations of genes.txt"))
```

```{r}
dc_nas_common_na <- dc_nas_common[!(dc_nas_common$gene_id %in% leg$gene_id),]
```

```{r}
neg <- genes_annotation$gene_name[genes_annotation$gene_id %in%
                             dc_nas_common_na$gene_id]
for (gene in neg) {
  print(gene)
}
```

```{r}
ss_df_short
```

```{r}
ss_coords$fiveSSs
```

```{r}
dc_nas_common
```
